smart-hadoop-support/smart-hadoop/src/main/java/org/smartdata/hdfs/scheduler/CopyScheduler.java

```diff
@@ -410,6 +418,60 @@ private void syncFileDiff() {
       }
     }
 
+    // add file_diff into Cache, if the file diff is already in cache, we update
+    // if the file diff is not in cache, we insert.
+    public void updateDiffIntoCache(List<FileDiff> fileDiffs) throws MetaStoreException {
+      LOG.debug("Add FileDiff Cache into file_diff cache");
+      for (FileDiff fileDiff : fileDiffs) {
+        fileDiffCache.put(fileDiff.getDiffId(), fileDiff);
+        syncTime++;
+      }
+      if (syncTime >= 50) {
+        syncTime = 0;
+        // update
+        pushCacheIntoDB();
+      }
+    }
+
+    // add file_diff into Cache, if the file diff is already in cache, we update
+    // if the file diff is not in cache, we insert.
+    public void updateDiffIntoCache(FileDiff fileDiff) {
+      LOG.debug("Add FileDiff Cache into file_diff cache");
+      fileDiffCache.put(fileDiff.getDiffId(), fileDiff);
+    }
+
+    public void deleteDiffInCache(Long did) {
+      LOG.debug("Delete FileDiff in cache");
+      if (fileDiffCache.containsKey(did)) {
+        fileDiffCache.remove(did);
+      }
+    }
+
+    public void updateFileDiffStatesInCache(Long did, FileDiffState fileDiffState) {
+      LOG.debug("Update FileDiff");
+      if (!fileDiffCache.containsKey(did)) {
+        return;
+      }
+      FileDiff fileDiff = fileDiffCache.get(did);
+      fileDiff.setState(fileDiffState);
+      //update
+      fileDiffCache.put(did, fileDiff);
+    }
+
+    public void pushCacheIntoDB() throws MetaStoreException {
```

Comment:
Push all cache to DB is not reasonable. Because file diff cache may be too large for one time sync.